### PR TITLE
fix: fail fast when training dataloaders yield zero batches

### LINKED ISSUE #921: When batch size is bigger than dataset size, the SFT finishes without error, but not doing any grad updates
**Describe the bug**

E.g. if bs=1024, but dataset only has 128 samples, I just see a bunch of these in the logs

========================= Epoch 1/2000 =========================

========================= Epoch 2/2000 =========================

========================= Epoch 3/2000 =========================

========================= Epoch 4/2000 =========================

========================= Epoch 5/2000 =========================

========================= Epoch 6/2000 =========================

========================= Epoch 7/2000 =========================

========================= Epoch 8/2000 =========================

and then the final checkpoint is saved but is the same as original.

**Steps/Code to reproduce bug**

Run any SFT with batch size bigger than dataset size

**Expected behavior**

Either fail with error or reduce batch size automatically to be equal to the dataset size.

### PR DESCRIPTION
**Pull request by tildahh**
# What does this PR do ?

Raises an actionable error during setup when a training dataloader would yield zero batches.

# Issues

Closes #921

# Usage

No configuration changes are required. For example, SFT with 128 samples and `policy.train_global_batch_size=1024` now raises an error directing the user to reduce the batch size or provide more data.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [x] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

With `drop_last=True`, an oversized batch can cause training to finish without any gradient updates. The shared guard covers SFT, DPO, RM, distillation, PPO, and GRPO, including GRPO’s per-task dataloaders. It reports the effective batch size, dataset size, and relevant configuration setting without automatically changing the batch size.

Repo pre-commit checks passed.

### CODE REVIEW

**Comment by tildahh** [comment]
Addressed in the latest commit. Updated the existing dataloader mocks with explicit lengths, extended the guard to the single-controller and cross-tokenizer distillation setups, and added focused tests for empty-dataloader rejection and one-batch acceptance.

Validation: 438 tests passed across seven unit-test suites, with no exclusions.

File: nemo_rl/algorithms/distillation.py [hunk 1]
```diff
@@ -299,6 +299,12 @@ def setup(
         collate_fn=rl_collate_fn,
         drop_last=True,
     )
+    check_train_dataloader_not_empty(
+        dataloader,
+        dataset=train_dataset,
+        batch_size=distillation_config.num_prompts_per_step,
+        batch_size_source="distillation.num_prompts_per_step",
+    )
```
[the remark this answers is not in the document]
Comment (tildahh) [reply nemo_rl/algorithms/distillation.py:307]: Fixed in 410ef18 by giving the affected dataloader mocks explicit positive lengths. The distillation suite now passes.

File: nemo_rl/algorithms/grpo.py [hunk 2]
```diff
@@ -701,6 +702,17 @@ def init_train_dataloader(dataset, suffix: str = ""):
...
+            dataloader,
+            dataset=dataset,
+            batch_size=dataloader_batch_size,
+            batch_size_source=(
+                "num_prompts_per_dataloader * batch_multiplier"
+                if data_config["use_multiple_dataloader"]
+                else "num_prompts_per_step * batch_multiplier"
+            ),
+            context=f" for task '{suffix.removeprefix('_')}'" if suffix else "",
+        )
```
[the remark this answers is not in the document]
Comment (tildahh) [reply nemo_rl/algorithms/grpo.py:715]: Fixed in 410ef18 by configuring the affected dataloader mocks to return a positive batch count. The GRPO suite now passes.
